barretenberg/cpp/src/barretenberg/ecc/groups/group.hpp

```diff
@@ -62,6 +65,82 @@ template <typename _coordinate_field, typename _subgroup_field, typename GroupPa
+     * significant bias when reducing modulo a ~256-bit coordinate_field NOTE: We ensure y-parity is linked to preimage
+     * hash because there is no canonical deterministic square root algorithm (i.e. if a field element has a square
+     * root, there are two of them and `field::sqrt` may return either one)
+     * @param num_generators
+     * @param domain_separator
+     * @return std::vector<affine_element>
+     */
+    inline static std::vector<affine_element> derive_generators_secure(const std::vector<uint8_t>& domain_separator,
+                                                                       const size_t num_generators,
+                                                                       const size_t starting_index = 0)
+    {
+        std::vector<affine_element> result;
+        std::array<uint8_t, 32> domain_hash = sha256::sha256(domain_separator);
+        std::vector<uint8_t> generator_preimage;
+        generator_preimage.reserve(64);
+        std::copy(domain_hash.begin(), domain_hash.end(), std::back_inserter(generator_preimage));
+        for (size_t i = 0; i < 32; ++i) {
+            generator_preimage.emplace_back(0);
+        }
+        for (size_t i = starting_index; i < starting_index + num_generators; ++i) {
+            auto generator_index = static_cast<uint32_t>(i);
+            uint32_t mask = 0xff;
+            generator_preimage[32] = static_cast<uint8_t>(generator_index >> 24);
+            generator_preimage[33] = static_cast<uint8_t>((generator_index >> 16) & mask);
+            generator_preimage[34] = static_cast<uint8_t>((generator_index >> 8) & mask);
+            generator_preimage[35] = static_cast<uint8_t>(generator_index & mask);
+            result.push_back(affine_element::hash_to_curve(generator_preimage));
+        }
+        return result;
+    }
+
+    inline static affine_element get_secure_generator_from_index(size_t generator_index,
+                                                                 const std::string& domain_separator)
+    {
+        std::array<uint8_t, 32> domain_hash = sha256::sha256(domain_separator);
+        std::vector<uint8_t> generator_preimage;
+        generator_preimage.reserve(64);
+        std::copy(domain_hash.begin(), domain_hash.end(), std::back_inserter(generator_preimage));
+        for (size_t i = 0; i < 32; ++i) {
+            generator_preimage.emplace_back(0);
+        }
+        auto gen_idx = static_cast<uint32_t>(generator_index);
+        uint32_t mask = 0xff;
+        generator_preimage[32] = static_cast<uint8_t>(gen_idx >> 24);
+        generator_preimage[33] = static_cast<uint8_t>((gen_idx >> 16) & mask);
+        generator_preimage[34] = static_cast<uint8_t>((gen_idx >> 8) & mask);
+        generator_preimage[35] = static_cast<uint8_t>(gen_idx & mask);
+        auto result = affine_element::hash_to_curve(generator_preimage);
+        ASSERT(result.x != 0);
+        ASSERT(result.y != 0);
```

Comment:
Noted above: I think we can get rid of this if we keep threading optional<affine_element> upto the caller instead of returning (0,0) -- I'm assuming this check is for checking for (0,0) which happens when the point is not on the curve

Reply:
I removed the ASSERTs as they were just debugging checks i forgot to remove. `hash_to_curve` should always return a point on the curve. It's a recursive algorithm that has a 50% chance of succeeding every call. So... 256 iterations and you've got a 1 / 2^{256} chance of not getting a point